node/network/availability-distribution/src/requester/fetch_task.rs

```diff
@@ -0,0 +1,397 @@
+		self.live_in.difference(leaves);
+		if self.live_in.is_empty() && !self.is_finished() {
+			self.state = FetchedState::Canceled
+		}
+	}
+
+	/// Whether or not there are still relay parents around with this candidate pending
+	/// availability.
+	pub fn is_live(&self) -> bool {
+		!self.live_in.is_empty()
+	}
+
+	/// Whether or not this task can be considered finished.
+	///
+	/// That is, it is either canceled, succeeded or failed.
+	pub fn is_finished(&self) -> bool {
+		match &self.state {
+			FetchedState::Canceled => true,
+			FetchedState::Started(sender) => sender.is_canceled(),
+		}
+	}
+}
+
+/// Things that can go wrong in task execution.
+#[derive(Debug)]
+enum TaskError {
+	/// The peer failed to deliver a correct chunk for some reason (has been reported as
+	/// appropriate).
+	PeerError,
+	/// This very node is seemingly shutting down (sending of message failed).
+	ShuttingDown,
+}
+
+impl RunningTask {
+	async fn run(self, kill: oneshot::Receiver<()>) {
+		// Wait for completion/or cancel.
+		let run_it = self.run_inner();
+		futures::pin_mut!(run_it);
+		let _ = select(run_it, kill).await;
+	}
+
+	/// Fetch and store chunk.
+	///
+	/// Try validators in backing group in order.
+	async fn run_inner(mut self) {
+		let mut bad_validators = Vec::new();
+		// Try validators in reverse order:
+		while let Some(validator) = self.group.pop() {
+			// Send request:
+			let resp = match self.do_request(&validator).await {
+				Ok(resp) => resp,
+				Err(TaskError::ShuttingDown) => {
+					tracing::info!(
+						target: LOG_TARGET,
+						"Node seems to be shutting down, canceling fetch task"
+					);
+					return
+				}
+				Err(TaskError::PeerError) => {
+					bad_validators.push(validator);
+					continue
+				}
+			};
+			let chunk = match resp {
+				AvailabilityFetchingResponse::Chunk(resp) => {
+					resp.reconstruct_erasure_chunk(&self.request)
+				}
+				AvailabilityFetchingResponse::NoSuchChunk => {
+					tracing::debug!(target: LOG_TARGET, validator = ?validator, "Validator did not have our chunk");
+					bad_validators.push(validator);
+					continue
+				}
+			};
+
+			// Data genuine?
+			if !self.validate_chunk(&validator, &chunk) {
+				bad_validators.push(validator);
+				continue;
+			}
+
+			// Ok, let's store it and be happy:
+			self.store_chunk(chunk).await;
+			break;
+		}
+		self.conclude(bad_validators).await;
+	}
+
+	/// Do request and return response, if successful.
+	async fn do_request(
+		&mut self,
+		validator: &AuthorityDiscoveryId,
+	) -> std::result::Result<AvailabilityFetchingResponse, TaskError> {
+		let (full_request, response_recv) =
+			OutgoingRequest::new(validator.clone(), self.request);
+		let requests = Requests::AvailabilityFetching(full_request);
+
+		self.sender
+			.send(FromFetchTask::Message(AllMessages::NetworkBridge(
+				NetworkBridgeMessage::SendRequests(vec![requests]),
+			)))
+			.await
+			.map_err(|_| TaskError::ShuttingDown)?;
+
+		match response_recv.await {
```

Comment:
What is the timeout here? I'd suggest parallelizing even at the cost of some bandwidth for now as we request randomly and only a simple majority are required to have the data.

Reply:
Just set it to three seconds. As long as we don't ensure an open connection via peer sets, we should probably not make this much lower.


> This PR already represents a huge improvement over the status quo and I would even be fine with requesting from all validators in the backing group at once (up to say, 10) and just taking the answer from whoever provides it first.

Ouch. What exactly is the required percentage of validators that must have the chunk? How is PoV distribution implemented? Do validators really stop, once a majority is reached? Or is it just, that, that's what is guaranteed, and under normal conditions all should have it? If it is the latter, I would lean towards being more conservative in parallel downloads.

Especially since none responding validators are ranked back anyway.

Reply:
@eskimor a simple majority (>1/2) should have it and that is all that's guaranteed. I expect that more will have it but I wouldn't place bets on more than 4/5 in typical conditions.

As mentioned elsewhere:

> Another thing to do would be to send notifications about having the data to all validators. And then validators can build up a list of other validators who claim to have the data rather than requesting from arbitrary others.

Reply:
However I still think that we're at the stage now where we want to prioritize availability over bandwidth. This PR is already a huge step forward in terms of bandwidth, so getting some redundant copies of our own chunk won't be a problem. We have clear strategies for optimization but I don't think we should spend too much time on this initial version as there are other subsystems to port over.

Reply:
Will not block on this; filed https://github.com/paritytech/polkadot/issues/2512